### DISCUSSION TITLE
Weird Code Process Threading Issue for ESP32-S3

### DISCUSSION THREAD

**Discussion by Type-32** [lvgl-micropython/lvgl_micropython#151] [Q&A] [answered]
Hardware: ESP32-S3-Touch-LCD-4.3
Code Repo: https://github.com/Type-32/GeneralPurposePDA
Replicate:
1. Flash in the LVGL Micropython firmware
2. flash the programs in the following order:
- globals.py
- pdaos_lib.py
- applications.py
- pdaos.py
- osui.py
- connectivity.py
- main.py
3. Press reset and you'll see UI on your screen
4. Click on the "Dices" app
5. Click on the input textarea
6. Type anything
7. Press the checkmark button on the keyboard, **the entire thing just stops running.**

It's a really weird issue with an easy workaround (adding a separate keyboard for that text area) but I just want to get this system up and going and use it a bit.

The related code is in **applications.py, lines 100~116**.
**The weird part about this is that, when you're at step 6, you enter the REPL, run the code under the `elif` statement, and everything works just fine without it being stuck.**
**_Pressing the button on the keyboard makes the entire thing stuck._** Which is weird as hell. I originally tried to do some debug output alongside and before the `elif` lines, and it's just stuck at the set_text() method, which, the code works fine if you run it separately in the REPL.

The other code works just fine.

The weird code part doesn't even output errors as well. I also tried running it asynchronously with an AsyncJob that constantly sets the input area's text value to the keyboard input area's text value, and nothing works. It just seems like it's clogged up somehow... no errors, no stack overflows, no memory leaks and stuff, just... stops.

If anyone knows anything on scheduling threads and design patterns, please, I am in need of some degree of help.

**Comment by Type-32** [comment, marked as answer]
The weird issue is fixed.

```py
        def DiceParserInput_eventhandler(event_struct):
            target = event_struct.get_target()
            event = event_struct.code
            if event == lv.EVENT.CLICKED:
                print("Editing dice parser text")
                # help(event_struct.get_target_obj())
                pdaos.invoke_keyboard(event_struct.get_target_obj().get_text())
                globals.set_temp_variable("ui_DiceParserInput", event_struct.get_target_obj())
                globals.set_temp_variable("diceParserFocused", True)
            elif globals.get_temp_variable("diceParserFocused") and not pdaos.is_keyboard_focused():
                def set_inputfield_text(*arg):
                    print("Woah")
                    # pdaos.revoke_keybaord()
                    globals.get_temp_variable("ui_DiceParserInput").set_text(pdaos.get_keyboard_text())
                    globals.remove_temp_variable("ui_DiceParserInput")

                globals.set_temp_variable("diceParserFocused", False)
                globals.remove_temp_variable("diceParserFocused")
                micropython.schedule(set_inputfield_text, None)
            return
```

Instead of calling it directly, using `micropython.schedule()` seems to fix the issue.

**Reply by kdschlosser** [reply]
There are some things you cannot do from a "thread" and can only be done from the "main thread" 

I am actually in the process of writing a more complete threading module. It is going to have most of the thing found in the CPython threading module. I am also adding a mutliprocessing module that will allow you to run stuff on the second core of the ESP32. YAY!!!

I am getting pretty close to being done with it.

**Reply by Type-32** [reply]
that's very cool 🤩 I look forward to using it <3

**Reply by Type-32** [reply]
Oh and one more question: how do I receive streaming responses with the `urequests` library, or is there another library I can use?
